astropy/table/operations.py

```diff
@@ -810,11 +815,11 @@ def _join(left, right, keys=None, join_type='inner',
             else:
                 # np.where does not work for mixin columns (e.g. Quantity) so
                 # use a slower workaround.
-                left_mask = ~right_mask
-                if np.any(left_mask):
-                    out[out_name][left_mask] = left[left_name].take(left_out)
+                non_right_mask = ~right_mask
                 if np.any(right_mask):
-                    out[out_name][right_mask] = right[right_name].take(right_out)
+                    out[out_name] = left[left_name].take(left_out)
```

Comment:
Why this change? And should this not be an in-place change at least? (with ``[:]``)

Reply:
indeed, it should use `[:]`. As for the rest of the changes, I did the fix to follow the very same logic as the `np.where` above (the issue with that it doesn't understand Quantities for np <1.17 and astropy <4.0):
```
                out[out_name][:] = np.where(right_mask,
                                            left[left_name].take(left_out),
                                            right[right_name].take(right_out))
```